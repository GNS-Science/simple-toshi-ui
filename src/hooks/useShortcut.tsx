import React from 'react';
import { useHistory } from 'react-router';

export function useShortcut(callback: () => void, keyCodes: string[], disableHotkey?: boolean): void {
  const history = useHistory();

  React.useEffect(() => {
    const handler = (event: KeyboardEvent) => {
      const searchParams = new URLSearchParams(history.location.search);
      if (keyCodes.includes(event.key.toLowerCase()) && !searchParams.get('modal')) {
        callback();
      }
    };
    if (!disableHotkey) {
      window.addEventListener('keydown', handler);
      return () => {
        window.removeEventListener('keydown', handler);
      };
    }
  }, [disableHotkey, keyCodes, history, callback]);
}
