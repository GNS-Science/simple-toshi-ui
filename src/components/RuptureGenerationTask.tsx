import React from 'react';
import { graphql } from 'babel-plugin-relay/macro';
import { Link, useParams } from 'react-router-dom';
import { RuptureGenerationTaskQuery } from './__generated__/RuptureGenerationTaskQuery.graphql';
import { useLazyLoadQuery } from 'react-relay/hooks';
import { Typography } from '@material-ui/core';
import KeyValueTable from './KeyValueTable';
import InfoTable from './InfoTable';
import FileTable from './FileTable';

const REFRESH_PERIOD = 3600000;

interface RuptureGenerationTaskParams {
  id: string;
}

const ruptureGenerationTaskQuery = graphql`
  query RuptureGenerationTaskQuery($id: ID!) {
    node(id: $id) {
      id
      ... on RuptureGenerationTask {
        id
        duration
        created
        result
        state
        files {
          edges {
            node {
              id
              role
              file {
                ... on File {
                  id
                  file_name
                  file_url
                }
              }
            }
          }
        }
        arguments {
          k
          v
        }
        environment {
          k
          v
        }
        metrics {
          k
          v
        }
        parents {
          edges {
            node {
              parent {
                ... on GeneralTask {
                  id
                }
              }
            }
          }
        }
      }
    }
  }
`;

const RuptureGenerationTask: React.FC = () => {
  const { id } = useParams<RuptureGenerationTaskParams>();
  const [fetchKey, setFetchKey] = React.useState<number>(0);

  React.useEffect(() => {
    const timeout = setTimeout(() => setFetchKey(fetchKey + 1), REFRESH_PERIOD);
    return () => {
      clearTimeout(timeout);
    };
  }, [fetchKey]);

  const data = useLazyLoadQuery<RuptureGenerationTaskQuery>(
    ruptureGenerationTaskQuery,
    { id },
    {
      fetchKey: fetchKey,
      fetchPolicy: 'network-only',
    },
  );

  if (!data?.node) {
    return (
      <Typography variant="h5" gutterBottom>
        Rupture Generation Task: Id Not Found
      </Typography>
    );
  }

  return (
    <>
      <Typography variant="h5" gutterBottom>
        Rupture Generation Task (id: {data?.node?.id})
      </Typography>
      <InfoTable
        created={data?.node?.created ? (data?.node?.created as string) : undefined}
        duration={data?.node?.duration}
        state={data?.node?.state}
        result={data?.node?.result}
      />
      <Typography>
        <strong>Parent tasks: </strong>
        {data?.node?.parents?.edges?.map((e, i, { length }) => {
          return (
            <React.Fragment key={e?.node?.parent?.id}>
              <Link to={`/GeneralTask/${e?.node?.parent?.id}`}>
                {Buffer.from(e?.node?.parent?.id ?? '', 'base64').toString()}
              </Link>
              {i + 1 !== length && <span>, </span>}
            </React.Fragment>
          );
        })}
      </Typography>
      {!!data?.node?.files?.edges?.length && <FileTable data={data?.node?.files?.edges} />}
      {data?.node?.arguments && <KeyValueTable header="Arguments" data={data?.node?.arguments} />}
      {data?.node?.environment && <KeyValueTable header="Environment" data={data?.node?.environment} />}
      {data?.node?.metrics && <KeyValueTable header="Metrics" data={data?.node?.metrics} />}
    </>
  );
};

export default RuptureGenerationTask;
