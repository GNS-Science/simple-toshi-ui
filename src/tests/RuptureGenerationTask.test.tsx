import { Suspense } from 'react';
import TestRenderer from 'react-test-renderer';
import { createMockEnvironment, MockPayloadGenerator, RelayMockEnvironment } from 'relay-test-utils';
import RuptureGenerationTask from '../components/RuptureGenerationTask';
import ReactRouter from 'react-router';
import { RelayEnvironmentProvider } from 'react-relay/hooks';

const mockResolver = {
  RuptureGenerationTask: () => ({
    id: '1234',
    duration: 60,
    created: '2021-05-19T07:04:42.283510+00:00',
    arguments: [
      {
        k: 'testArgument',
        v: 'testArgumentValue',
      },
    ],
    metrics: [
      {
        k: 'testMetric',
        v: 'testMetricValue',
      },
    ],
    environment: [
      {
        k: 'testEnvironment',
        v: 'testEnvironmentValue',
      },
    ],
    result: 'SUCCESS',
    state: 'DONE',
  }),
};

const setup = (environment: RelayMockEnvironment) => {
  return TestRenderer.create(
    <RelayEnvironmentProvider environment={environment}>
      <Suspense fallback="Loading...">
        <RuptureGenerationTask />
      </Suspense>
    </RelayEnvironmentProvider>,
  );
};

describe('RuptureGenerationTask component', () => {
  it('displays a RuptureGenerationTask using mock graphql payload', () => {
    jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ id: '1234' });
    const environment = createMockEnvironment();
    environment.mock.queueOperationResolver((operation) => MockPayloadGenerator.generate(operation, mockResolver));

    const renderer = setup(environment);

    expect(renderer).toMatchSnapshot();
    renderer.unmount();
  });

  it('displays not found when no matching id', () => {
    jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ id: '1234' });
    const environment = createMockEnvironment();
    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        RuptureGenerationTask() {
          return null;
        },
      }),
    );

    const renderer = setup(environment);

    expect(renderer).toMatchSnapshot();
    renderer.unmount();
  });

  it('calls graphql query with url param', () => {
    jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ id: '1234' });
    const environment = createMockEnvironment();

    const renderer = setup(environment);

    const variables = environment.mock.getMostRecentOperation().request.variables;
    expect(variables).toStrictEqual({ id: '1234' });
    renderer.unmount();
  });
});
