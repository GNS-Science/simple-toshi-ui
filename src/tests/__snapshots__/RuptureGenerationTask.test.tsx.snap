// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuptureGenerationTask component displays a RuptureGenerationTask using mock graphql payload 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthMd"
  style={
    Object {
      "paddingTop": "40px",
      "wordWrap": "break-word",
    }
  }
>
  <h5
    className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
  >
    Rupture Generation Task (id: 
    1234
    )
  </h5>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <ul
        className="MuiList-root MuiList-dense MuiList-padding"
      >
        <li
          className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
          disabled={false}
        >
          <div
            className="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
            >
              Created
            </span>
            <p
              className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              2021-05-19T07:04:42.283510+00:00
            </p>
          </div>
        </li>
        <li
          className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
          disabled={false}
        >
          <div
            className="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
            >
              Duration
            </span>
            <p
              className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              60
            </p>
          </div>
        </li>
      </ul>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <ul
        className="MuiList-root MuiList-dense MuiList-padding"
      >
        <li
          className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
          disabled={false}
        >
          <div
            className="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
            >
              Result
            </span>
            <p
              className="MuiTypography-root MuiListItemText-secondary makeStyles-success-1 MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              SUCCESS
            </p>
          </div>
        </li>
        <li
          className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
          disabled={false}
        >
          <div
            className="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
            >
              State
            </span>
            <p
              className="MuiTypography-root MuiListItemText-secondary makeStyles-success-1 MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              DONE
            </p>
          </div>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="MuiPaper-root makeStyles-root-4 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      className="MuiTable-root makeStyles-table-5 MuiTable-stickyHeader"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-7 MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
            colSpan={2}
            scope="col"
          >
            Arguments
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-7"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-6 MuiTableCell-sizeSmall"
          >
            testArgument
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-6 MuiTableCell-sizeSmall"
          >
            testArgumentValue
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="MuiPaper-root makeStyles-root-4 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      className="MuiTable-root makeStyles-table-5 MuiTable-stickyHeader"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-7 MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
            colSpan={2}
            scope="col"
          >
            Environment
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-7"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-6 MuiTableCell-sizeSmall"
          >
            testEnvironment
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-6 MuiTableCell-sizeSmall"
          >
            testEnvironmentValue
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="MuiPaper-root makeStyles-root-4 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      className="MuiTable-root makeStyles-table-5 MuiTable-stickyHeader"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-7 MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
            colSpan={2}
            scope="col"
          >
            Metrics
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-7"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-6 MuiTableCell-sizeSmall"
          >
            testMetric
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableCell-6 MuiTableCell-sizeSmall"
          >
            testMetricValue
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`RuptureGenerationTask component displays not found when no matching id 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthMd"
  style={
    Object {
      "paddingTop": "40px",
      "wordWrap": "break-word",
    }
  }
>
  <h5
    className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
  >
    Rupture Generation Task (id: 
    Not found
    )
  </h5>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <ul
        className="MuiList-root MuiList-dense MuiList-padding"
      >
        <li
          className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
          disabled={false}
        >
          <div
            className="MuiListItemText-root MuiListItemText-dense"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
            >
              Created
            </span>
          </div>
        </li>
        <li
          className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
          disabled={false}
        >
          <div
            className="MuiListItemText-root MuiListItemText-dense"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
            >
              Duration
            </span>
          </div>
        </li>
      </ul>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <ul
        className="MuiList-root MuiList-dense MuiList-padding"
      >
        <li
          className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
          disabled={false}
        >
          <div
            className="MuiListItemText-root MuiListItemText-dense"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
            >
              Result
            </span>
          </div>
        </li>
        <li
          className="MuiListItem-root MuiListItem-dense MuiListItem-gutters"
          disabled={false}
        >
          <div
            className="MuiListItemText-root MuiListItemText-dense"
          >
            <span
              className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
            >
              State
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="MuiPaper-root makeStyles-root-11 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      className="MuiTable-root makeStyles-table-12 MuiTable-stickyHeader"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-14 MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
            colSpan={2}
            scope="col"
          >
            Arguments
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
  <div
    className="MuiPaper-root makeStyles-root-11 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      className="MuiTable-root makeStyles-table-12 MuiTable-stickyHeader"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-14 MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
            colSpan={2}
            scope="col"
          >
            Environment
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
  <div
    className="MuiPaper-root makeStyles-root-11 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      className="MuiTable-root makeStyles-table-12 MuiTable-stickyHeader"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-14 MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
            colSpan={2}
            scope="col"
          >
            Metrics
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
</div>
`;
